### PR TITLE
feat(eventual-send): breakpoint on delivery by env-options

### PR DESCRIPTION
**Pull request by erights**
closes: #XXXX
refs: #XXXX

## Description

Given an environment variable setting of

```console
$ export ENDO_DELIVERY_BREAKPOINTS="quatloos issuer.getAmountOf=*"
```

then when `E` invokes a `getAmountOf` method on an object with @@toStringTag of `'Alleged: quatloos issuer'`, it should first execute a JavaScript `debugger;` statement. When run normally, i.e., not under a debugger, the `debugger;` statement has no effect. But when run under a debugger, this statement acts as a breakpoint, causing execution to pause there.

If the "*" at the end is replaced by a number, then each time execution matches, the number is decremented by 1. When it reaches zero, it executes `debugger;`.

If the "getAmountOf" is replaced with "*", then it breakpoints on any method of a quatloos issuer.

If the "quatloos issuer" is replaced with "*", then it breakpoints on calling a method named "getAmountOf" on anything. 

Similarly

```console
$ export ENDO_SEND_BREAKPOINTS="quatloos issuer.getAmountOf=*"
```

will breakpoint when the message is eventually sent, so that you can look up the stack in the debugger to understand why it was sent. The name only matches if the object itself, or a remote presence for the object, is the target and so has the correct @@toStringTag. This does not work when eventually sending to a promise. To catch those cases, use "Promise" or "*" for the target name.

### Security Considerations

As with all use of env-options, we should worry about whether an adversary can read or write these options, and so interfere with code that is supposed to use these options.

### Scaling Considerations

If the `ENDO_DELIVERY_BREAKPOINTS` env-option is not even set, this implementation takes a fast path which should have essentially no penalty compared to the absence of these mechanisms.

When `ENDO_DELIVERY_BREAKPOINTS` is present but has no "*" on the method name, we still try to minimize the overhead on non-matching method names. But once we have a method name match, then we do a linear search among these directives on each such call.

Likewise for `ENDO_SEND_BREAKPOINTS`

### Documentation Considerations

Let's get some road experience ourselves first. But if it proves useful, it'll be a great debugging aid to explain.

It will be interesting to figure out how to show how to use it interactively from inside a debugger well, including dynamically changing breakpoints. Probably a short video?

### Testing Considerations

Since this alters only the behavior seen from a debugger, it is unclear how to write an automated test.

### Upgrade Considerations

None.

### CODE REVIEW

**Comment by dckc** [comment]
This looks like a new product feature. I'd like us to bring @sufyaankhan up to speed and get his take on it.

> Documentation Considerations
> Let's get some road experience ourselves first.

Before merging it? Or before documenting it?

**Comment by erights** [comment]
> This looks like a new product feature. I'd like us to bring @sufyaankhan up to speed and get his take on it.

Absolutely! @sufyaankhan would love your thoughts.

> 
> > Documentation Considerations
> > Let's get some road experience ourselves first.
> 
> Before merging it? Or before documenting it?

Before documenting. We won't get enough road experience ourselves if we don't merge it first.

File: packages/eventual-send/src/local.js
Comment (dckc) [inline packages/eventual-send/src/local.js:41]: `harden(BREAKS)` after this for loop?

**Review by dckc — COMMENTED** [review]
I looked it over. I don't see any critical problems, but I haven't studied `makeEnvironmentCaptor` well enough to be comfortable with this yet. Nor have I tested it manually.

**Review by turadg — COMMENTED** [review]
Since this is a tentative name/feature, please indicate that. E.g. `EXPERIMENTAL_BREAKPOINTS`. That makes clear that it isn't supported and may break.

When/if it is advanced from experimentation, the name should be scoped to what it does. It's not a way to express any breakpoint, just ones for remote calls. E.g. `E_CALL_BREAKPOINTS`. Or for more harmony with the extant `DEBUG=` option, `DEBUG_E_CALLS`.

File: packages/eventual-send/src/local.js
Comment (gibson042) [inline packages/eventual-send/src/local.js:41]: Grouping primarily by method name/wildcard in appearance order and secondarily by object tag/wildcard in appearance order implies a particular specificity resolution such that e.g. `BREAKPOINTS=Alleged: Foo.*=1,Alleged: Foo.m=2,*.m=3,Alleged: Bar.m=4,*.m2=5` will result in `E(fooInstance).m()` using the 2 while `E(barInstance).m()` and `E(bazInstance).m()` use the 3 (which in particular completely supersedes the 4), and `E(fooInstance).m2()` using the 5. Prioritizing method name over object tag seems fine, but we probably want wildcards to be considered last in each group. For example:
```suggestion
    // eslint-disable-next-line no-nested-ternary
    const count = countStr === '*'
      ? countStr
      : countStr === '0'
        ? 0
        : (+countStr || Fail`invalid skip count in delivery breakpoint filter ${q(bp)}`);
    const patts = hasOwn(BREAKS, verb) ? BREAKS[verb] : (BREAKS[verb] = []);
    const i = tag === '*' ? -1 : patts.findIndex(patt => patt[0] === '*');
    if (i === -1) {
      patts.push([tag, count]);
    } else {
      patts.splice(i, 0, [tag, count]);
    }
  }
}
if (BREAKS['*']) {
  const wildcard = BREAKS['*'];
  delete BREAKS['*'];
  BREAKS['*'] = wildcard;
}
```

**Review by gibson042 — COMMENTED** [review]
I think the concept has merit, but I agree that we'll want some experience with it to really decide.

File: packages/eventual-send/src/local.js
Comment (gibson042) [inline packages/eventual-send/src/local.js:67]: Is there a benefit to this `Record<Verb, Array<[Tag, Count]>>` structure over `Record<Verb, Record<Tag, Count>>`?
```suggestion
        const countsByTag = BREAKS[verb] || BREAKS['*'];
        if (countsByTag !== undefined) {
          const objTag = obj[Symbol.toStringTag];
          const matchTag = hasOwn(countsByTag, objTag) ? objTag : '*';
          const count = countsByTag[matchTag];
          if (count === '*' || count === 0) {
            // eslint-disable-next-line no-debugger
            debugger;
          } else if (count !== undefined) {
            countsByTag[matchTag] = count - 1;
          }
        }
        return apply(f, obj, args);
```

File: packages/eventual-send/src/local.js
Comment (michaelfig) [reply packages/eventual-send/src/local.js:21]: I appreciate @turadg's concern over the naming of the `BREAKPOINTS` environment variable.  I think as long as it is prefixed with `ENDO_`, we can use whatever name we want without fear of confusion or stomping on somebody else's variables.

```suggestion
const BREAKPOINTS = getEnvironmentOption('ENDO_E_BREAKPOINTS', '');
```

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: Better, I've removed the extra function call from the stack. Rather, I replaced it with a predicate call 
```js
if (shouldBreakpointDelivery(...) {
  breakpoint;
}
``` 
immediately before the real apply, which also creates a much more pleasant experience in the debugger.

As part of the refactoring, I did also give all these better names. Resolving.

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: The rename to `ENDO_DELIVERY_BREAKPOINTS` is done. This also made it easy to add `ENDO_SEND_BREAKPOINTS` for an additional feature which may be even more useful.

Not resolving this conversation yet, as I haven't yet made the suggested representation change

**Comment by erights** [comment]
I have not yet made the suggested representation changes, so I reverted this PR to Draft.

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: I took @michaelfig suggestion below to just go with a JSON string as the value of the option, completely avoiding the need for regexp matching. Resolving.

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: The data structure previously name `BREAKS` is now the encapsulated `breakpointsTable`. Its internal representation still needs reform, but this has not yet happened. With the current representation, I countdown by mutating the part containing the `count` number in place. So, not yet resolving, but noting to clarify further review.

Comment (erights) [reply packages/eventual-send/src/local.js]: @gibson042 this internal representation is now encapsulated but is not yet reformed. So, not yet resolving, but noting to clarify further review.

@dckc , I refactored the decision about whether to breakpoint into a separate predicate function that does not itself do the breakpointing, but instead leaves that to its caller. As a result, I was able to write unit tests testing that predicate. PTAL.

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: I have made this suggested external representation change. But I have not yet made the suggested internal one. Nor have I yet done anything to support symbol-named methods. I like the idea of using a manually implemented tiny subset of smallcaps for this one case, so I will examine that.

Thus, not yet resolving.

**Comment by erights** [comment]
Note to reviewers: Enough changed that it wasn't worth preserving old commits, so I squashed since you last reviewed.

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: Since the internal breakpoints table is encapsulated, I didn't freeze so I could continue to do this trick.

Separately, to support interactive change of breakpoints, I now make the external representation available from within the debugger. To ease modification, I separately do not freeze it, so it can be modified in place.

@dckc since your suggestion ended in a "?", I'll consider this suggestion done. But please unresolve this conversation if you disagree.

Comment (erights) [reply packages/eventual-send/src/local.js]: With the new nested-record internal representation, I was more easily able to consistently always check the '*' cases last. Resolving.

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: Done

File: packages/eventual-send/src/local.js
Comment (erights) [reply packages/eventual-send/src/local.js]: Shift to internal representation done.

I have not yet done anything about symbol-named methods. But I don't think that needs to happen in this PR. Resolving.

**Comment by erights** [comment]
Reviewers,
Now that this is ready for review, I'll avoid further squashes without first asking you.

File: packages/eventual-send/src/E.js
Comment (michaelfig) [inline packages/eventual-send/src/E.js:133]: Any reason this `HandledPromise...` call appears *before* the `onSend` checking?  All the other calls to `HandledPromise` static methods appear to be *after* the `onSend`.

File: packages/eventual-send/src/message-breakpoints.js
Comment (michaelfig) [inline packages/eventual-send/src/message-breakpoints.js:13]: Maybe document in terms of types, like.

```js
/** @typedef {string | '*'} MatchStringTag docs... */
/** @typedef {string | '*'} MatchMethodName docs... */
/** @typedef {number | '*'} MatchCountdown docs... */ 
```

```suggestion
 * @typedef {Record<MatchStringTag, Record<MatchMethodName, MatchCountdown>>} MessageBreakpoints
```

File: packages/eventual-send/src/message-breakpoints.js
Comment (michaelfig) [inline packages/eventual-send/src/message-breakpoints.js:22]:
```suggestion
 * @typedef {Record<MatchMethodName, Record<MatchStringTag, MatchCountdown>>} BreakpointTable
```

File: packages/eventual-send/src/message-breakpoints.js
Comment (michaelfig) [inline packages/eventual-send/src/message-breakpoints.js:47]: Does it ever happen that a string tag has both prefixes?  If not, can you please document some examples of them in the jsdoc above the function?

File: packages/eventual-send/src/message-breakpoints.js
Comment (michaelfig) [inline packages/eventual-send/src/message-breakpoints.js:111]: I think this should be:

```suggestion
    if (methodName === undefined || methodName === null) {
```

File: packages/eventual-send/src/message-breakpoints.js
Comment (michaelfig) [reply packages/eventual-send/src/message-breakpoints.js:13]: Also, could you please describe explicitly what `MatchCountdown` does?  It is not obvious from reading the other jsdocs whether this is the number of noops until entering the debugger, or the number of debugger traps until it becomes a noop.

File: packages/eventual-send/src/E.js
Comment (erights) [reply packages/eventual-send/src/E.js]: My mistake. Thanks for catching! Done.

File: packages/eventual-send/src/message-breakpoints.js
Comment (erights) [reply packages/eventual-send/src/message-breakpoints.js]: Done.

File: packages/eventual-send/src/message-breakpoints.js
Comment (erights) [reply packages/eventual-send/src/message-breakpoints.js]: Done

File: packages/eventual-send/src/message-breakpoints.js
Comment (erights) [reply packages/eventual-send/src/message-breakpoints.js:78]: Done

File: packages/eventual-send/src/message-breakpoints.js
Comment (erights) [reply packages/eventual-send/src/message-breakpoints.js]: Done

File: packages/eventual-send/src/message-breakpoints.js
Comment (michaelfig) [inline packages/eventual-send/src/message-breakpoints.js:10]:
```suggestion
 *   matched against the value of a recipient's `@@toStringTag`
```

File: packages/eventual-send/src/message-breakpoints.js
Comment (michaelfig) [inline packages/eventual-send/src/message-breakpoints.js:25]:
```suggestion
 *   must be a non-negative integer. Once that is zero, always breakpoint.
```

**Review by michaelfig — APPROVED** [review]
Some trivial doc fixes, and you may want a second opinion before merging, but LGTM!

File: packages/eventual-send/src/message-breakpoints.js
Comment (erights) [reply packages/eventual-send/src/message-breakpoints.js]: Done

File: packages/eventual-send/src/message-breakpoints.js
Comment (erights) [reply packages/eventual-send/src/message-breakpoints.js]: Done

**Pull request merged**
